streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -192,29 +191,113 @@ public void processInOrder(final K key, final V value, final long timestamp) {
             //create left window for new record
             if (!leftWinAlreadyCreated) {
                 final ValueAndTimestamp<Agg> valueAndTime;
-                //there's a right window that the new record could create --> new record's left window is not empty
-                if (latestLeftTypeWindow != null) {
+                // if there's a right window that the new record could create && previous record falls within left window -> new record's left window is not empty
+                if (previousRecord != null && leftWindowNotEmpty(previousRecord, timestamp)) {
                     valueAndTime = ValueAndTimestamp.make(leftWinAgg.value(), timestamp);
                 } else {
                     valueAndTime = ValueAndTimestamp.make(initializer.apply(), timestamp);
                 }
                 final TimeWindow window = new TimeWindow(timestamp - windows.timeDifferenceMs(), timestamp);
                 putAndForward(window, valueAndTime, key, value, closeTime, timestamp);
             }
-            //create right window for new record
             if (!rightWinAlreadyCreated && rightWindowIsNotEmpty(rightWinAgg, timestamp)) {
-                final TimeWindow window = new TimeWindow(timestamp + 1, timestamp + 1 + windows.timeDifferenceMs());
-                final ValueAndTimestamp<Agg> valueAndTime = ValueAndTimestamp.make(getValueOrNull(rightWinAgg), Math.max(rightWinAgg.timestamp(), timestamp));
+                createRightWindow(timestamp, rightWinAgg, key, value, closeTime);
+            }
+        }
+
+        /**
+         * Created to handle records that have a timestamp > 0 but < timeDifference. These records would create
+         * windows with negative start times, which is not supported. Instead, they will fall within the [0, timeDifference]
+         * window, and we will update their right windows as new records come in later
+         */
+        private void processEarly(final K key, final V value, final long timestamp, final long closeTime) {
+            ValueAndTimestamp<Agg> rightWinAgg = null;
+            //window from [0,timeDifference] that holds all early records
```

Comment:
```suggestion
            // A window from [0, timeDifferenceMs] that holds all early records
```
Also I'd suggest putting the `combinedWindow` declaration (and comment) above `rightWinAgg` to avoid ambiguity in what the comment refers to